conda/env/env.py

```diff
@@ -207,7 +219,400 @@ def add(self, package_name):
         self.parse()
 
 
-class Environment:
+@total_ordering
+class Requirement:
+    """A single requirement for a package.
+
+    Can either be a single spec string (e.g. "flask"), or a dictionary containing
+    a conditionally present package, and a spec to include if that condition is present.
+    For example:
+
+    {
+        "if": "__linux",    # <-- if __linux is present, also include flask as a requirement
+        "then": "flask",
+    }
+    """
+
+    def __init__(self, spec: str | dict[str, str]):
+        if isinstance(spec, str):
+            self.spec = spec
+            self.condition = None
+        elif isinstance(spec, dict):
+            for key in ["if", "then"]:
+                if key not in spec:
+                    raise ValueError(f"Conditional requirement missing an `{key}` key.")
+
+            self.spec = spec["then"]
+            self.condition = spec["if"]
+        else:
+            raise ValueError(f"Invalid requirement: {spec}")
+
+    def __repr__(self) -> str:
+        result = self.spec
+        if self.condition:
+            result = f"{result} (if {self.condition})"
+        return result
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, Requirement):
+            return self.spec == other.spec
+
+        raise NotImplementedError
+
+    def __lt__(self, other: Requirement) -> bool:
+        return self.spec < other.spec
+
+    def serialize(self) -> str | dict[str, str]:
+        """Serialize the package requirement.
+
+        Can be a conditional requirement of the form
+
+        {
+            "if": "__win",
+            "then": "flask"
+        }
+
+        :return: Serialized Requirement
+        """
+        if self.condition:
+            return {"if": self.condition, "then": self.spec}
+        return self.spec
+
+    def __str__(self) -> str:
+        return repr(self)
+
+
+class Requirements:
+    """A set of conda and PyPI requirements for a package."""
```

Comment:
I don't like having these two aspects combined in a single class 🤔 They should be separate imo. Any particular reason why you merged them here?

Reply:
Because it makes sense conceptually - a package has a set of requirements. Those requirements can be either conda or pypi requirements or both. Maybe this shouldn't be a separate class, and instead each `Requirement` can be either a `PyPIRequirement` or a `CondaRequirement`? I'll try that approach now.